apis/compute/v1beta1/computefirewallpolicyrule_reference.go

```diff
@@ -0,0 +1,166 @@
+	// Should be in the format "locations/global/firewallPolicies/<firewallPolicy>/rules/<priority>".
+	External string `json:"external,omitempty"`
+
+	// The name of a ComputeFirewallPolicyRule resource.
+	Name string `json:"name,omitempty"`
+
+	// The namespace of a ComputeFirewallPolicyRule resource.
+	Namespace string `json:"namespace,omitempty"`
+
+	parent *ComputeFirewallPolicyRuleParent
+}
+
+// NormalizedExternal provision the "External" value for other resource that depends on ComputeFirewallPolicyRule.
+// If the "External" is given in the other resource's spec.ComputeFirewallPolicyRuleRef, the given value will be used.
+// Otherwise, the "Name" and "Namespace" will be used to query the actual ComputeFirewallPolicyRule object from the cluster.
+func (r *ComputeFirewallPolicyRuleRef) NormalizedExternal(ctx context.Context, reader client.Reader, otherNamespace string) (string, error) {
+	if r.External != "" && r.Name != "" {
+		return "", fmt.Errorf("cannot specify both name and external on %s reference", ComputeFirewallPolicyRuleGVK.Kind)
+	}
+	// From given External
+	if r.External != "" {
+		if _, _, err := parseComputeFirewallPolicyRuleExternal(r.External); err != nil {
+			return "", err
+		}
+		return r.External, nil
+	}
+
+	// From the Config Connector object
+	if r.Namespace == "" {
+		r.Namespace = otherNamespace
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(ComputeFirewallPolicyRuleGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return "", fmt.Errorf("reading referenced %s %s: %w", ComputeFirewallPolicyRuleGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return "", fmt.Errorf("ComputeFirewallPolicyRule is not ready yet")
+	}
+	r.External = actualExternalRef
+	return r.External, nil
+}
+
+// New builds a NewComputeFirewallPolicyRuleRef from the Config Connector ComputeFirewallPolicyRule object.
+func NewComputeFirewallPolicyRuleRef(ctx context.Context, reader client.Reader, obj *ComputeFirewallPolicyRule) (*ComputeFirewallPolicyRuleRef, error) {
+	id := &ComputeFirewallPolicyRuleRef{}
+
+	firewallPolicyRef, err := refsv1beta1.ResolveComputeFirewallPolicy(ctx, reader, obj, obj.Spec.FirewallPolicyRef)
+	if err != nil {
+		return nil, err
+	}
+	firewallPolicy := firewallPolicyRef.External
+	if firewallPolicy == "" {
+		return nil, fmt.Errorf("cannot resolve firewallPolicy")
+	}
+
+	id.parent = &ComputeFirewallPolicyRuleParent{FirewallPolicy: firewallPolicy}
+
+	// Get priority. Priority is a required field
+	priority := fmt.Sprintf("%v", obj.Spec.Priority)
```

Comment:
I think I used `priority` as `resourceID` here. User provide an integer and I converted it to string here.
The reason is `resourceID` is not a supported field for this resource: https://cloud.google.com/config-connector/docs/reference/resource-docs/compute/computefirewallpolicyrule, and we may not want to change the beta CRD for now